### PR TITLE
Add rewind to chain processor

### PR DESCRIPTION
**Pull request by danield9tqh**
## Summary
ChainProcessor was failing to update correctly if the head was behind the chain head

## Testing Plan
Unit tests and local testing

## Breaking Change

Is this a breaking change? If yes, add notes below on why this is breaking and
what additional work is required, if any.

```
[ ] Yes
```

### CODE REVIEW

File: ironfish/src/chainProcessor.ts [hunk 1]
```diff
@@ -96,20 +96,40 @@ export class ChainProcessor {
       }
     }
 
-    const iter = this.chain.iterateTo(fork, chainHead, undefined, false)
+    if (chainHead.sequence < head.sequence) {
```
Comment (NullSoldier) [inline ironfish/src/chainProcessor.ts]: Why cant you just change the `isLinear` above to this condition?

Comment (hughy) [reply ironfish/src/chainProcessor.ts]: `chainHead` could still be ahead of `head` and you'd need to do the block above. Suppose `chainHead` is chain A and `head` is chain B:
```
A1 -> A2 -> A3 -> A4
               -> B3
```
`head` needs to go backwards from B3 -> A2, and then forwards from A2 -> A3 -> A4

Comment (hughy) [reply ironfish/src/chainProcessor.ts]: I think we might be able to change it to `if (fork.sequence < head.sequence)`. The block below seems to assume `fork === head`

Comment (NullSoldier) [reply ironfish/src/chainProcessor.ts]: Sorry, yes. That's what I mean. fork.sequence < head.sequence

File: ironfish/src/chainProcessor.ts [hunk 2]
```diff
@@ -73,32 +73,30 @@ export class ChainProcessor {
...
-        await this.remove(remove)
-        this.hash = remove.previousBlockHash
-        this.sequence = remove.sequence - 1
+      if (remove.hash.equals(fork.hash)) {
+        continue
       }
+
+      await this.remove(remove)
+      this.hash = remove.hash
+      this.sequence = remove.sequence
```
Comment (hughy) [inline ironfish/src/chainProcessor.ts:94]: Should these be setting it to the block before remove?

File: ironfish/src/chainProcessor.ts (hunk 1 above)
Comment (danield9tqh) [reply ironfish/src/chainProcessor.ts]: Hugh was saying we can always just rewind to the fork then fast-forward to the destination and that should handle all cases. Seems sound. Added a comment here for explanation

File: ironfish/src/chainProcessor.ts (hunk 2 above)
Comment (danield9tqh) [reply ironfish/src/chainProcessor.ts]: Updated

**Review by hughy — APPROVED** [review]

File: ironfish/src/chainProcessor.ts [hunk 3]
```diff
@@ -73,32 +73,30 @@ export class ChainProcessor {
       `Chain processor head not found in chain: ${this.hash.toString('hex')}`,
     )
 
-    const { fork, isLinear } = await this.chain.findFork(head, chainHead)
-    if (!fork) {
-      return { hashChanged: false }
-    }
+    const { fork } = await this.chain.findFork(head, chainHead)
```
Comment (NullSoldier) [inline ironfish/src/chainProcessor.ts:76]: What is the fork here in the case that chainHead is ahead? This is ok?

Comment (danield9tqh) [reply ironfish/src/chainProcessor.ts:76]: Fork will return the smaller of the two if they are on the same linear branch

**Pull request merged**